File: vox-rpg/src/rpg.ts
export type PlayerStats = {
  level: number;
  xp: number;
  xpToNext: number;
  health: number;
  maxHealth: number;
  stamina: number;
  maxStamina: number;
  mana: number;
  maxMana: number;
};

export type Inventory = {
  selectedIndex: number;
  slots: { id: string; count: number }[];
};

export type SaveState = {
  stats: PlayerStats;
  inventory: Inventory;
  position: { x: number; y: number; z: number };
};

const STORAGE_KEY = 'vox_rpg_save_v1';

export function createInitialStats(): PlayerStats {
  return {
    level: 1,
    xp: 0,
    xpToNext: 50,
    health: 100,
    maxHealth: 100,
    stamina: 100,
    maxStamina: 100,
    mana: 50,
    maxMana: 50,
  };
}

export function addXp(stats: PlayerStats, amount: number): void {
  stats.xp += amount;
  while (stats.xp >= stats.xpToNext) {
    stats.xp -= stats.xpToNext;
    stats.level += 1;
    // Scale next level requirement and increase stats slightly
    stats.xpToNext = Math.floor(stats.xpToNext * 1.35 + 25);
    stats.maxHealth += 10;
    stats.maxStamina += 10;
    stats.maxMana += 5;
    stats.health = stats.maxHealth;
    stats.stamina = stats.maxStamina;
    stats.mana = Math.min(stats.maxMana, stats.mana + 10);
  }
}

export function regenTick(stats: PlayerStats, dt: number): void {
  // Stamina regen
  stats.stamina = Math.min(stats.maxStamina, stats.stamina + 8 * dt);
  // Mana regen
  stats.mana = Math.min(stats.maxMana, stats.mana + 3 * dt);
}

export function saveGame(state: SaveState): void {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(state));
  } catch {}
}

export function loadGame(): SaveState | null {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return null;
    return JSON.parse(raw) as SaveState;
  } catch {
    return null;
  }
}

File: vox-rpg-py/rpg.py
from dataclasses import dataclass, asdict
from typing import Dict, Any
import json
import pathlib

SAVE_PATH = pathlib.Path(__file__).resolve().parent / 'save.json'


@dataclass
class PlayerStats:
    level: int = 1
    xp: int = 0
    xp_to_next: int = 50
    health: float = 100
    max_health: float = 100
    stamina: float = 100
    max_stamina: float = 100
    mana: float = 50
    max_mana: float = 50

    def add_xp(self, amount: int):
        self.xp += amount
        while self.xp >= self.xp_to_next:
            self.xp -= self.xp_to_next
            self.level += 1
            self.xp_to_next = int(self.xp_to_next * 1.35 + 25)
            self.max_health += 10
            self.max_stamina += 10
            self.max_mana += 5
            self.health = self.max_health
            self.stamina = self.max_stamina
            self.mana = min(self.max_mana, self.mana + 10)

    def regen(self, dt: float):
        self.stamina = min(self.max_stamina, self.stamina + 8 * dt)
        self.mana = min(self.max_mana, self.mana + 3 * dt)


@dataclass
class Inventory:
    selected_index: int = 0
    slots: Any = None

    def __post_init__(self):
        if self.slots is None:
            self.slots = [
                { 'id': 'grass', 'count': 999 },
                { 'id': 'dirt', 'count': 999 },
                { 'id': 'stone', 'count': 999 },
                { 'id': 'log', 'count': 999 },
                { 'id': 'plank', 'count': 999 },
            ]


def save_game(data: Dict[str, Any]):
    try:
        with open(SAVE_PATH, 'w') as f:
            json.dump(data, f)
    except Exception as e:
        print('Failed to save game:', e)


def load_game() -> Dict[str, Any]:
    try:
        if SAVE_PATH.exists():
            with open(SAVE_PATH, 'r') as f:
                return json.load(f)
    except Exception as e:
        print('Failed to load game:', e)
    return {}